boards/feather_m0/examples/adalogger.rs

```diff
@@ -0,0 +1,185 @@
+#![no_std]
+#![no_main]
+
+use feather_m0 as hal;
+use panic_halt as _;
+
+use core::fmt::Write;
+use core::sync::atomic;
+
+use cortex_m::interrupt::free as disable_interrupts;
+use cortex_m::peripheral::NVIC;
+use embedded_sdmmc::{Controller, SdMmcSpi, VolumeIdx};
+use usb_device::bus::UsbBusAllocator;
+use usb_device::prelude::*;
+use usbd_serial::{SerialPort, USB_CLASS_CDC};
+
+use hal::clock::{ClockGenId, ClockSource, GenericClockController};
+use hal::delay::Delay;
+use hal::entry;
+use hal::pac::{interrupt, CorePeripherals, Peripherals};
+use hal::prelude::*;
+use hal::rtc;
+use hal::time::U32Ext;
+use hal::usb::UsbBus;
+
+use heapless::consts::U1024;
+use heapless::String;
+
+#[entry]
+fn main() -> ! {
+    // setup basic peripherals
+    let mut peripherals = Peripherals::take().unwrap();
+    let mut core = CorePeripherals::take().unwrap();
+    let mut clocks = GenericClockController::with_internal_32kosc(
+        peripherals.GCLK,
+        &mut peripherals.PM,
+        &mut peripherals.SYSCTRL,
+        &mut peripherals.NVMCTRL,
+    );
+    let mut delay = Delay::new(core.SYST, &mut clocks);
+
+    // configure the peripherals we'll need
+    // get the internal 32k running at 1024 Hz for the RTC
+    let timer_clock = clocks
+        .configure_gclk_divider_and_source(ClockGenId::GCLK3, 32, ClockSource::OSC32K, true)
+        .unwrap();
+    let rtc_clock = clocks.rtc(&timer_clock).unwrap();
+    let timer = rtc::Rtc::clock_mode(peripherals.RTC, rtc_clock.freq(), &mut peripherals.PM);
+    let mut pins = hal::Pins::new(peripherals.PORT);
+    let mut red_led = pins.d13.into_open_drain_output(&mut pins.port);
+
+    let bus_allocator = unsafe {
+        USB_ALLOCATOR = Some(hal::usb_allocator(
+            peripherals.USB,
+            &mut clocks,
+            &mut peripherals.PM,
+            pins.usb_dm,
+            pins.usb_dp,
+        ));
+        USB_ALLOCATOR.as_ref().unwrap()
+    };
+
+    unsafe {
+        USB_SERIAL = Some(SerialPort::new(&bus_allocator));
+        USB_BUS = Some(
+            UsbDeviceBuilder::new(&bus_allocator, UsbVidPid(0x16c0, 0x27dd))
+                .manufacturer("Fake company")
+                .product("Serial port")
+                .serial_number("TEST")
+                .device_class(USB_CLASS_CDC)
+                .build(),
+        );
+    }
+
+    unsafe {
+        core.NVIC.set_priority(interrupt::USB, 1);
+        NVIC::unmask(interrupt::USB);
+    }
+
+    // Now work on the SD peripherals. Slow SPI speed required on init
+    let spi = hal::spi_master(
+        &mut clocks,
+        400_u32.khz(),
```

Comment:
LGTM, but maybe this is a good opportunity to demonstrate the spi reconfiguring in #427 . The low speed is only needed for sd card init, and can be reconfigured after to a higher speed.

Reply:
I agree, I'd like to incorporate that. Will wait for that to merge, then will make the change.

Reply:
FYI, #427 is only for `sercom::v2::spi`. To use it, you would have to update your SPI. If you're willing, I would encourage you to try it. So far I think I'm the only one using it. Actually, I guess @jbeaurivage must be using it as well. But doing so may be more work than you're interested in right now.

Reply:
Ah screw it, I'm gonna try it along with GPIO v2. If I can do it this weekend lets just go for it

Reply:
Alright, trying to use v2 SPI @bradleyharden and hitting:
```rust
/// Convenience for setting up the labelled SPI peripheral.
/// This powers up SERCOM4 and configures it for use as an
/// SPI Master in SPI Mode 0.
pub fn spi_master<F: Into<Hertz>>(
    // clocks: &mut GenericClockController,
    bus_speed: F,
    sercom4: pac::SERCOM4,
    pm: &mut pac::PM,
    pins: &mut v2::Pins,
) -> spi::Spi<spi::Config<
        spi::Pads<pac::SERCOM4,
        Pad<pac::SERCOM4, Pad0, v2::PA12>,
        Pad<pac::SERCOM4, Pad2, v2::PB10>,
        Pad<pac::SERCOM4, Pad3, v2::PB11>>>> {
    let pads = spi::Pads::<Sercom4>::new()
        .sclk::<Pad3, _>(pins.pb11)
        .data_in::<Pad0, _>(pins.pa12)
        .data_out::<Pad2, _>(pins.pb10);

    spi::Config::new(&pm, sercom4, pads, bus_speed).enable()
}
```

I'm hitting this on all of the pins:
```rust
error[E0507]: cannot move out of `pins.pb11` which is behind a mutable reference
   --> src/lib.rs:283:26
    |
283 |         .sclk::<Pad3, _>(pins.pb11)
    |                          ^^^^^^^^^ move occurs because `pins.pb11` has type `hal::gpio::v2::Pin<PB11, hal::gpio::v2::Disabled<hal::gpio::v2::Floating>>`, which does not implement the `Copy` trait
```

Is there a way to get a pin reference without like all pins moving into the spi struct? I'm following the documentation in `hal/src/common/thumbv6m/sercom/v2/spi.rs`, so if there's a better way I can tweak those docs.

Reply:
@TDHolmes, check out the `pads_alias` macro. I think that's what it's called. I can't remember the name off the top of my head. It would make it easier for you to define the return type.

Unfortunately, I don't think there's any solution to your ownership problem. The `Spi` struct needs to own the corresponding pins. Doesn't the same problem exist for the `v1` module? I'm not sure how best to handle that. Perhaps look at some other BSPs? I think the wio_terminal had a good structure to it.

Reply:
I'm also available on the Matrix channel now, for more synchronous debugging.

Reply:
ohhh duh that's why v1 passed individual pins & by value lol. I think this'll work. I'll be working on this on & off today and probably tomorrow. sounds like I should finally get Matrix :p

Reply:
For reference: the [`pads_alias`](https://atsamd-rs.github.io/docs/samd21g/thumbv6m-none-eabi/doc/atsamd_hal/macro.pads_alias.html) macro.